File: 백준/Silver/9012. 괄호/괄호.java
import java.util.Scanner;
import java.util.Stack;

public class Main {
    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);

        int t = sc.nextInt();

        for (int i = 0; i < t; i++) {
            System.out.println(stack(sc.next()));
        }
    }

    public static String stack(String s) {

        Stack<Character> st = new Stack<>();

        for (int i = 0; i < s.length(); i++) {

            char c = s.charAt(i);

            if (c == '(') {
                st.push(c);
            }

            else if (st.empty()) {
                return "NO";
            }

            else {
                st.pop();
            }
        }

        if (st.empty()) {
            return "YES";
        } else {
            return "NO";
        }
    }
}

File: 백준/Silver/9012. 괄호/괄호.py
t = int(input())

def bracket_stack(bracket):
    stack = []
    for s in bracket:
        if s == '(':
            stack.append('(')
        elif s == ')':
            if len(stack) == 0:
                return "NO"
            else:
                stack.pop()
    if len(stack) == 0:
        return "YES"
    else:
        return "NO"

for _ in range(t):
    bracket = input()
    print(bracket_stack(bracket))